FLT/Deformations/Categories.lean

```diff
@@ -0,0 +1,230 @@
+import FLT.Deformations.IsProartinian
+import FLT.Deformations.IsResidueAlgebra
+import Mathlib.Topology.Algebra.Algebra.Equiv
+
+universe u
+
+open CategoryTheory Function Limits IsLocalRing
+
+namespace Deformation
+
+variable (𝓞 : Type u) [CommRing 𝓞]
+
+/-- A local proartinian algebra is a topological ring that is local and proartinian,
+and that the induced map on the residue fields is bijective. -/
+class IsLocalProartinianAlgebra
+    (R : Type u) [CommRing R] [TopologicalSpace R] [Algebra 𝓞 R] : Prop extends
+  IsTopologicalRing R, IsLocalRing R, IsProartinian R,
+  IsLocalHom (algebraMap 𝓞 R), IsResidueAlgebra 𝓞 R
+
+/-- The category of local proartinian algebras over `𝓞` with fixed residue field `𝕜`. -/
+structure ProartinianCat where
+  /-- Underlying set of a `ProartinianCat` -/
+  carrier : Type u
+  /-- ring structure of a `ProartinianCat` -/
+  [commRing : CommRing carrier]
+  /-- topological space structure of a `ProartinianCat` -/
+  [topologicalSpace : TopologicalSpace carrier]
+  /-- algebra structure of a `ProartinianCat` -/
+  [algebra : Algebra 𝓞 carrier]
+  [isLocalProartinianAlgebra : IsLocalProartinianAlgebra 𝓞 carrier]
+
+local notation3:max "𝓒" 𝓞 => ProartinianCat 𝓞
+
+namespace ProartinianCat
+
+attribute [instance] commRing algebra topologicalSpace isLocalProartinianAlgebra
+
+initialize_simps_projections ProartinianCat (-commRing, -algebra, -topologicalSpace)
+
+instance : CoeSort (ProartinianCat 𝓞) (Type u) := ⟨carrier⟩
+
+attribute [coe] ProartinianCat.carrier
+
+/-- Make a `ProartinianCat` from an unbundled algebra. -/
+abbrev of (X : Type u) [CommRing X] [Algebra 𝓞 X]
+    [TopologicalSpace X] [IsLocalProartinianAlgebra 𝓞 X] :
+    ProartinianCat 𝓞 :=
+  ⟨X⟩
+
+lemma coe_of (X : Type u) [CommRing X] [Algebra 𝓞 X] [TopologicalSpace X]
+    [IsLocalProartinianAlgebra 𝓞 X] :
+    of 𝓞 X = X := rfl
+
+variable {𝓞} in
+/-- The type of morphisms in `ProartinianCat 𝓞`. -/
+@[ext]
+structure Hom (A B : ProartinianCat 𝓞) where
+  /-- The underlying algebra map. -/
+  hom : A →A[𝓞] B
```

Comment:
Note to self: are any of "continuous", "local" and "O-algebra map" implied by the others?